src/execution/sql/runtime_types.cpp

```diff
@@ -655,4 +658,711 @@ Timestamp Timestamp::FromYMDHMSMU(int32_t year, int32_t month, int32_t day, int3
+  value_ = 0;
+
+  if (input.empty()) return;
+
+  uint32_t pos = 0;
+
+  bool is_negative = false;
+  if (input[pos] == '-') {
+    pos++;
+    is_negative = true;
+  }
+
+  while (pos < input.size() && input[pos] != '.') {
+    value_ += input[pos] - '0';
+    value_ *= 10;
+    pos++;
+  }
+
+  if (precision == 0) {
+    value_ /= 10;
+    if (pos != input.size()) {
+      if (pos + 1 < input.size()) {
+        pos++;
+        if (input[pos] - '0' > 5) {
+          value_ += 1;
+        } else if (input[pos] - '0' == 5 && value_ % 2 == 1) {
+          value_ += 1;
+        }
+      }
+    }
+    if (is_negative) {
+      value_ = -value_;
+    }
+    return;
+  }
+
+  // No decimal point case
+  if (pos == input.size()) {
+    for (uint32_t i = 0; i < precision - 1; i++) {
+      value_ *= 10;
+    }
+    if (is_negative) {
+      value_ = -value_;
+    }
+    return;
+  }
+  // Skip decimal point
+  pos++;
+  // Nothing after decimal point case
+  if (pos == input.size()) {
+    for (uint32_t i = 0; i < precision - 1; i++) {
+      value_ *= 10;
+    }
+    if (is_negative) {
+      value_ = -value_;
+    }
+    return;
+  }
+
+  for (uint32_t i = 1; i < precision; i++) {
+    if (pos < input.size()) {
+      value_ += input[pos] - '0';
+      value_ *= 10;
+      pos++;
+    } else {
+      for (uint32_t j = i; j < precision; j++) {
+        value_ *= 10;
+      }
+      if (is_negative) {
+        value_ = -value_;
+      }
+      return;
+    }
+  }
+
+  if (pos == input.size()) {
+    if (is_negative) {
+      value_ = -value_;
+    }
+    return;
+  }
+
+  if (pos == input.size() - 1) {
+    // No Rounding required
+    value_ += input[pos] - '0';
+  } else {
+    if (input[pos + 1] - '0' > 5) {
+      // Round Up
+      value_ += input[pos] - '0' + 1;
+    } else if (input[pos + 1] - '0' < 5) {
+      // No Rounding will happen
+      value_ += input[pos] - '0';
+    } else {
+      if ((input[pos] - '0') % 2 == 0) {
+        // Round up if ODD
+        value_ += input[pos] - '0';
+      } else {
+        // Round up if ODD
+        value_ += input[pos] - '0' + 1;
+      }
+    }
+  }
+
+  if (is_negative) {
+    value_ = -value_;
+  }
+}
+
+void CalculateMultiWordProduct128(const uint128_t *const half_words_a, const uint128_t *const half_words_b,
+                                  uint128_t *half_words_result, uint32_t m, uint32_t n) {
+  uint128_t k, t;
+  uint32_t i, j;
+  constexpr const uint128_t bottom_mask = (uint128_t{1} << 64) - 1;
+  for (i = 0; i < m; i++) half_words_result[i] = 0;
+  for (j = 0; j < n; j++) {
+    k = 0;
+    for (i = 0; i < m; i++) {
+      t = half_words_a[i] * half_words_b[j] + half_words_result[i + j] + k;
+      half_words_result[i + j] = t & bottom_mask;
+      k = t >> 64;
+    }
+    half_words_result[j + m] = k;
+  }
+}
+
+int Nlz128(uint128_t x) {
+  // Not used Figure 5-19 - Hacker's Delight double method as we need for 128 bits
```

Comment:
Not used??

Reply:
Used in tests

Reply:
Oh sorry. Yes some optimizations not used